src/gt4py/next/iterator/transforms/infer_domain.py

```diff
@@ -0,0 +1,145 @@
+        if shift:
+            off, val = shift
+            current_dim = cls._infer_dimension_from_offset(off)
+            new_ranges[current_dim.value] = SymbolicRange.translate(
+                symbolic_domain.ranges[current_dim.value], val.value
+            )
+        return SymbolicDomain("cartesian_domain", new_ranges)
+
+    @classmethod
+    def infer_as_fieldop(
+            cls,
+            stencil: itir.FunCall,
+            input_domain: SymbolicDomain
+    ) -> Tuple[itir.FunCall, Dict[str, SymbolicDomain]]: # todo: test scan operator
+
+        assert isinstance(stencil, itir.FunCall) and isinstance(stencil.fun, itir.FunCall)
+        assert stencil.fun.fun == im.ref('as_fieldop')
+
+        inputs = stencil.args
+        tmp_counter = 0
+        inputs_node = inputs.copy()
+
+        # Set inputs for StencilClosure node by replacing FunCalls with temporary SymRefs
+        for idx, in_field in enumerate(inputs):
+            if isinstance(in_field, itir.FunCall):
+                in_field.id = im.ref(f'__dom_inf_{tmp_counter}')
+                inputs_node[idx] = im.ref(in_field.id)
+                tmp_counter += 1
+
+        accessed_domains = {str(in_field.id): [] for in_field in inputs}
+        out_field_name = "tmp"  # todo: can this be derived from somewhere?
+
+        symbolic_domain = cls._get_symbolic_domain(input_domain)
+
+        # Build node that can be passed to TraceShifts
+        node = itir.StencilClosure(
+            stencil=stencil.fun.args[0],
+            inputs=inputs_node,
+            output=im.ref(out_field_name),
+            domain=SymbolicDomain.as_expr(symbolic_domain),
+        )
+
+        # Extract the shifts and translate the domains accordingly
+        for in_field in inputs:
+            in_field_id = str(in_field.id)
+            shifts_list = TraceShifts.apply(node)[in_field_id]
+            dims = cls._extract_axis_dims(SymbolicDomain.as_expr(symbolic_domain))
+
+            new_domains = [
+                cls._translate_domain(symbolic_domain, shift, dims)
+                for shift in shifts_list
+            ]
+
+            accessed_domains[in_field_id] = domain_union(new_domains)
+
+        inputs_new = inputs.copy()
+        for idx, in_field in enumerate(inputs):
+            # Recursively traverse inputs
+            if isinstance(in_field, itir.FunCall):
+                transformed_calls_tmp, accessed_domains_tmp = cls.infer_as_fieldop(in_field,
+                                                                                   accessed_domains[str(in_field.id)])
+                inputs_new[idx] = transformed_calls_tmp
+
+                # Merge accessed_domains and accessed_domains_tmp
+                for k, v in accessed_domains_tmp.items():
+                    if k in accessed_domains:
+                        accessed_domains[k] = domain_union([accessed_domains[k], v])
+                    else:
+                        accessed_domains[k] = v
```

Comment:
I think this belongs to the beginning of this function. This will also cleanup the awkward multiple copies of the inputs (or I am overlooking something...).